### PR TITLE
PASTIS: add data module, SITS compatibility

### PR DESCRIPTION
**Pull request by yichiac**
Related to https://github.com/microsoft/torchgeo/issues/2382
Depends on multiclass time-series semantic segmentation support

- [x]  PASTIS: add new data module
- [x]  SemanticSegmentationTask: change to support T x C x H x W and C x H x W

### CODE REVIEW

File: torchgeo/datamodules/__init__.py [hunk 1]
```diff
@@ -34,6 +34,7 @@
 from .naip import NAIPChesapeakeDataModule
 from .nasa_marine_debris import NASAMarineDebrisDataModule
 from .oscd import OSCDDataModule
+from .pastis import PastisDataModule
```
Comment (adamjstewart) [inline torchgeo/datamodules/__init__.py]: Needs to be added to `__all__`

File: tests/conf/pastis.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    backbone: 'resnet18'
+    in_channels: 24
+    task: 'binary'
+    in_channels: 12
+    loss: 'bce'
+data:
+  class_path: PastisDataModule
+  init_args:
+    batch_size: 2
+    val_split_pct: 0.5
```
Comment (adamjstewart) [inline tests/conf/pastis.yaml]: This is not (yet) a valid argument for the `PastisDataModule`

File: tests/conf/pastis.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+model:
+  class_path: SemanticSegmentationTask
+  init_args:
+    model: 'unet'
+    backbone: 'resnet18'
+    in_channels: 24
```
Comment (adamjstewart) [inline tests/conf/pastis.yaml]: `in_channels` is duplicated

File: torchgeo/datamodules/pastis.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+"""PASTIS datamodule."""
+
+from typing import Any
+
+import kornia.augmentation as K
+
+from ..datasets import PASTIS
+from .geo import NonGeoDataModule
+
+class PastisDataModule(NonGeoDataModule):
```
Comment (adamjstewart) [inline torchgeo/datamodules/pastis.py]: Since the dataset is PASTIS, I think the data module should be PASTISDataModule

File: torchgeo/datamodules/pastis.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+        Args:
+            stage: Either 'fit', 'validate', 'test', or 'predict'.
+        """
+        if stage in ['fit']:
+            self.train_dataset = PASTIS(split='train', **self.kwargs)
+        if stage in ['fit', 'validate']:
+            self.val_dataset = PASTIS(split='val', **self.kwargs)
+        if stage in ['predict']:
+            self.predict_dataset = PASTIS(split='test', **self.kwargs)
```
Comment (adamjstewart) [inline torchgeo/datamodules/pastis.py]: PASTIS does not have pre-defined splits, you need to create your own

File: torchgeo/trainers/segmentation.py [hunk 6]
```diff
@@ -100,6 +104,20 @@ class and used with 'ce' loss.
         self.weights = weights
         super().__init__()
 
+    def forward(self, x:Tensor) -> Tensor:
```
Comment (adamjstewart) [inline torchgeo/trainers/segmentation.py]: While this is a clever solution, it only works for the currently available models. We also want to add time-series models like Satlas, Conv-LSTM, CNN-LSTM, etc. Satlas is already available and could be added to this PR, the others are coming soon.

File: torchgeo/trainers/segmentation.py [hunk 7]
```diff
@@ -100,6 +104,20 @@ class and used with 'ce' loss.
...
+        """
+        Args:
+            x: Input tensor of shape (B, C, H, W) or (B, T, C, H, W).
+
+        Returns:
+            Output tensor of shape (B, T*C, H, W).
```
Comment (adamjstewart) [inline torchgeo/trainers/segmentation.py]: I don't think these output dimensions are correct, isn't it B x "# classes" x H x W?

**Comment by yichiac** [comment]
Thanks, @adamjstewart. I will work on:
- [x] create the train-val-test splits for pastis.

File: docs/api/datamodules.rst [hunk 8]
```diff
@@ -175,6 +175,11 @@ OSCD
 
 .. autoclass:: OSCDDataModule
 
+PASTIS
+^^^^^^^^^^
```
Comment (adamjstewart) [inline docs/api/datamodules.rst]:
```suggestion
^^^^^^
```

File: torchgeo/trainers/segmentation.py [hunk 9]
```diff
@@ -100,6 +104,19 @@ class and used with 'ce' loss.
...
+
+        Args:
+            x: Input tensor of shape (B, C, H, W) or (B, T, C, H, W).
+
+        Returns:
+            Output tensor of shape (B, num_classes, H, W).
+        """
+        if x.ndim == 5:
+            x = rearrange(x, 'b t c h w -> b (t c) h w')
+        return cast(Tensor, self.model(x))
```
Comment (adamjstewart) [inline torchgeo/trainers/segmentation.py]:
```suggestion
        x = self.model(x)
        return x
```

File: tests/conf/pastis.yaml [hunk 10]
```diff
@@ -0,0 +1,15 @@
+model:
+  class_path: SemanticSegmentationTask
+  init_args:
+    model: 'unet'
+    backbone: 'resnet18'
+    task: 'binary'
+    in_channels: 12
```
Comment (adamjstewart) [inline tests/conf/pastis.yaml]: `in_channels: 12` is correct, but I think you need to tell the model that it's actually `in_channels * T`

File: tests/conf/pastis.yaml [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    loss: 'bce'
+data:
+  class_path: PASTISDataModule
+  init_args:
+    batch_size: 2
+    val_split_pct: 0.5
```
Comment (adamjstewart) [inline tests/conf/pastis.yaml]:
```suggestion
    val_split_pct: 0.33
    test_split_pct: 0.33
```
or something like that

File: tests/conf/pastis.yaml [hunk 12]
```diff
@@ -0,0 +1,15 @@
+model:
+  class_path: SemanticSegmentationTask
+  init_args:
+    model: 'unet'
+    backbone: 'resnet18'
+    task: 'binary'
```
Comment (adamjstewart) [inline tests/conf/pastis.yaml]: I don't think PASTIS is binary, there are 18 + 1 classes, right?

File: tests/conf/pastis.yaml [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+    in_channels: 12
+    loss: 'bce'
```
Comment (adamjstewart) [inline tests/conf/pastis.yaml]:
```suggestion
    loss: 'ce'
```

File: tests/conf/pastis.yaml (hunk 12 above)
Comment (yichiac) [reply tests/conf/pastis.yaml]: Yes, it's not binary. Multi-label seems to be the right one.

Comment (adamjstewart) [reply tests/conf/pastis.yaml]: I don't think it's multilabel either, should be multiclass...

Comment (yichiac) [reply tests/conf/pastis.yaml]: You are right, it's multiclass.

File: tests/conf/pastis.yaml (hunk 10 above)
Comment (yichiac) [reply tests/conf/pastis.yaml]: How do we define T in this case?

File: tests/conf/pastis.yaml (hunk 12 above)
Comment (yichiac) [inline tests/conf/pastis.yaml]:
```suggestion
    task: 'multiclass'
```

File: tests/conf/pastis.yaml (hunk 10 above)
Comment (adamjstewart) [reply tests/conf/pastis.yaml]: Hmm, I guess it won't actually be easy to know this in advance... Let's just use `in_channels: C * T`. So basically, expect the user to do the math.

Comment (adamjstewart) [reply tests/conf/pastis.yaml]: @keves1 @hkristen this will result in a slight API difference between `SemanticSegmentationTask` and `ChangeDetectionTask`, where the former will use `in_channels: C x T` and the latter will use `in_channels: C` where C is implicitly multiplied by 2 internally. Should we change `ChangeDetectionTask` to match `SemanticSegmentationTask`? Or make one a subclass of the other?

File: tests/conf/pastis.yaml [hunk 14]
```diff
@@ -0,0 +1,16 @@
+model:
+  class_path: SemanticSegmentationTask
+  init_args:
+    model: 'unet'
+    backbone: 'resnet18'
+    task: 'multiclass'
+    in_channels: C * T
```
Comment (adamjstewart) [inline tests/conf/pastis.yaml]: This needs to be a number

**Comment by yichiac** [comment]
The remaining issue is caused by Kornia's augmentation. It might be caused by the changing time steps for different modalities (S1A, S1D, S2). I am thinking adding customized `on_after_batch_transfer` to datamodule to pad or truncate to the target time steps. But this would limit the flexibility.

**Comment by adamjstewart** [comment]
One of the TODOs is to add a padding utility for this purpose. But this use case seems particularly complicated since you have to pad 3 different things. Can you confirm that the 3 different band subsets have different length, not just the entire collection? Also, what order are they in? S1A_T1, S1D_T1, S2_T1, S1A_T2, ... or S1A_T1, S1A_T2, ..., S1D_T1?

**Comment by yichiac** [comment]
Each band subset has a different length itself. I checked the time stamps (first element of data shapes) for all modalities. Here are the results:
```
S2
61: 626 times
46: 738 times
43: 537 times
42: 4 times
38: 562 times
37: 1 times

S1A
71: 1819 times
70: 108 times
65: 541 times

S1D
70: 1730 times
69: 738 times
```

**Comment by yichiac** [comment]
The previous error was caused by varying max timestamps across train/val/test splits, which led to mismatched input channel sizes when passing data to the unet. My solution is to pad all samples to a consistent maximum timestamp, ensuring stable input shapes (B, 61, 10, 128, 128) across all splits for S2.

For S2, the global maximum timestamp is 61 based on the previous survey. It is also the second most frequent case. This makes it highly likely that each batch has at least one sample with 61 timestamps, so almost all batches already align to shape (B, 61, 10, 128, 128). Thus, the padding size is set to 61 timestamps for all batches to ensure consistent input shapes across all splits.

File: torchgeo/datamodules/pastis.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+import kornia.augmentation as K
+import torch
+from torch import Tensor
+from torch.utils.data import random_split
+
+from ..datasets import PASTIS
+from .geo import NonGeoDataModule
+
+
+def collate_fn(
```
Comment (adamjstewart) [inline torchgeo/datamodules/pastis.py]: This should be in `torchgeo/datasets/utils.py` with the other collation functions so we can easily reuse it across multiple datamodules.

File: torchgeo/datamodules/pastis.py [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+    batch: list[dict[str, Tensor]], max_timestamp: int = 61
+) -> dict[str, Any]:
+    """Custom timeseries collate fn to handle variable length sequences.
+
+    Args:
+        batch: list of sample dicts return by dataset
+        max_timestamp: maximum length of the time series
```
Comment (adamjstewart) [inline torchgeo/datamodules/pastis.py]: I want to pad each mini-batch to the max length of that mini-batch. This will result in each mini-batch having a different length. This means PASTIS won't be compatible with U-Net, it would only be compatible with models supporting a dynamic length sequence.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: torchgeo/datamodules/pastis.py (hunk 15 above)
Comment (yichiac) [reply torchgeo/datamodules/pastis.py]: I will move this function to `utils.py`.

Comment (yichiac) [reply torchgeo/datamodules/pastis.py]: I moved the function to `torchgeo/datamodules/utils.py`

**Comment by yichiac** [comment]
The current version uses the max length of the entire dataset, since most mini-batches are already padded to this length. If we still want dynamic lengths, we need models that support them.

File: torchgeo/datamodules/utils.py [hunk 17]
```diff
@@ -43,6 +43,44 @@ def collate_fn_detection(batch: list[dict[str, Tensor]]) -> dict[str, Any]:
     return output
 
 
+def collate_fn_timeseries(
```
Comment (adamjstewart) [inline torchgeo/datamodules/utils.py]: * Move to torchgeo/_datasets_/utils.py (nearby stack_samples, etc.)
* Rename to temporal_collate, temporal_pad (check what other PyTorch padding functions are called)
* Add a `pad_length: int | None = None` parameter (if None, pad to max length of mini-batch, else pad to `pad_length`) pad_length or max_length is fine
* Have to truncate, either clearly document this or issue a warning

File: tests/trainers/test_segmentation.py [hunk 18]
```diff
@@ -117,6 +118,9 @@ def test_trainer(
             '1',
         ]
 
+        if name == 'pastis':
+            args.extend(['--data.max_timestamp', '9'])
```
Comment (adamjstewart) [inline tests/trainers/test_segmentation.py]: Put this in `pastis.yaml`?

**Comment by yichiac** [comment]
- [x] Move to torchgeo/datasets/utils.py (nearby stack_samples, etc.)
- [x] Rename to temporal_collate, temporal_pad (check what other PyTorch padding functions are called) -> `pad_sequence` 
- [x] Add a pad_length: int | None = None parameter (if None, pad to max length of mini-batch, else pad to pad_length) pad_length or max_length is fine
- [x] Have to truncate, either clearly document this or issue a warning

**Comment by adamjstewart** [comment]
`pad_sequence` is missing a lot of coverage. Just to be sure, did you try the default `pad_sequence` from PyTorch to make sure it doesn't work for us?

**Comment by yichiac** [comment]
I haven't experimented with the default `pad_sequence`. Will try it on our test data.

**Comment by yichiac** [comment]
The efficiency is improved after using the default `pad_sequence` and `pad` from PyTorch. The time required to loop through all batches is reduced from `63.26` seconds to `42.46` seconds.

File: tests/conf/pastis.yaml [hunk 19]
```diff
@@ -0,0 +1,18 @@
+model:
+  class_path: SemanticSegmentationTask
+  init_args:
+    model: 'unet'
+    backbone: 'resnet18'
+    task: 'multiclass'
+    in_channels: 610 # C * T (61 timesteps * 10 S2 channels)
```
Comment (adamjstewart) [inline tests/conf/pastis.yaml]: Shorten to max length of test data, not real data

File: torchgeo/datasets/utils.py [hunk 20]
```diff
@@ -421,6 +424,54 @@ def _dict_list_to_list_dict(
     return uncollated
 
 
+def pad_across_batches(
+    batch: list[dict[str, Tensor]],
+    padding_value: float = 0.0,
+    padding_length: int | None = None,
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py]: Thought about this more and it seems like `None` is just a wrapper around `pad_sequence`. Let's remove support for `None` and people can instead use `pad_sequence` directly if that's the behavior they want.

File: torchgeo/datasets/utils.py [hunk 21]
```diff
@@ -421,6 +424,54 @@ def _dict_list_to_list_dict(
...
+) -> dict[str, Any]:
+    """Custom timeseries collate fn to handle variable length sequences.
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py]:
```suggestion
    """Custom time-series collate fn to handle variable length sequences.
```

File: torchgeo/trainers/segmentation.py [hunk 22]
```diff
@@ -296,12 +314,26 @@ def validation_step(
...
         ):
             datamodule = self.trainer.datamodule
-            aug = K.AugmentationSequential(
-                K.Denormalize(datamodule.mean, datamodule.std),
-                data_keys=None,
-                keepdim=True,
-            )
-            batch = aug(batch)
+            if batch['image'].ndim == 5:
+                B, T, C, H, W = batch['image'].shape
```
Comment (adamjstewart) [inline torchgeo/trainers/segmentation.py]:
```suggestion
                _, T, C, _, _ = batch['image'].shape
```

File: tests/datasets/test_utils.py [hunk 23]
```diff
@@ -537,3 +538,38 @@ def test_azcopy(tmp_path: Path, azcopy: Executable) -> None:
...
+
+
+def test_pad_across_batches() -> None:
+    batch = [
+        {'image': torch.ones(2, 10, 5, 5), 'mask': torch.zeros(5, 5)},
+        {'image': torch.ones(3, 10, 5, 5), 'mask': torch.zeros(5, 5)},
+    ]
+
+    with pytest.raises(ValueError, match='padding_length must be specified'):
+        pad_across_batches(batch, padding_value=0.0, padding_length=cast(int, None))
```
Comment (adamjstewart) [inline tests/datasets/test_utils.py]: None is not an int, you can remove this test

File: torchgeo/datasets/utils.py [hunk 24]
```diff
@@ -421,6 +422,61 @@ def _dict_list_to_list_dict(
...
+    Returns:
+        batch dict output
+
+    .. versionadded:: 0.8
+    """
+    if padding_length is None:
+        raise ValueError(
+            'padding_length must be specified.'
+            'If pad to the max length in the batch, use torch.nn.utils.rnn.pad_sequence directly.'
+        )
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py]: Let's remove this and rely on the type hints to guide people

File: torchgeo/datasets/utils.py [hunk 25]
```diff
@@ -421,6 +422,61 @@ def _dict_list_to_list_dict(
...
 
 
+def pad_across_batches(
+    batch: list[dict[str, Tensor]], padding_length: int, padding_value: float = 0.0
+) -> dict[str, Any]:
+    """Custom time-series collate fn to handle variable length sequences.
+
+    Args:
+        batch: list of sample dicts returned by dataset
+        padding_length: the length to pad the sequences
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py]:
```suggestion
        padding_length: the length to pad the sequences to
```
Otherwise it sounds like the amount we add, which varies

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Last remaining changes, then we merge and move on to the next dataset!
